### PR TITLE
`Development`: Add multiple cards as user interface components for the new sidebar

### PR DESCRIPTION
**Pull request by egekurt123**
#### General

- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/#naming-conventions-for-github-pull-requests).

#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data.
- [x] I **strictly** followed the [client coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [x] I documented the TypeScript code using JSDoc style.

### Description

This PR introduces cards of varying sizes "small", "medium", and "large" that are designed to deliver content tailored to the specific needs of different modules, such as lectures, exercises, and conversations. Additionally, a directive has been implemented to facilitate the generation of these cards

**Note:** This PR only contains the card components. The application of these cards in specific sidebar configurations will be addressed in following PRs. "The reason this PR is separated from its usage is that these cards will be utilized by other developers in the near future."

#### Code Review
- [x] Code Review 1
- [x] Code Review 2
- [x] Code Review 3
- [x] Code Review 4

### Test Coverage
| Class/File | Line Coverage 
|------------|--------------:|
| [sidebar-card.directive](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6554-22/artifact/shared/Coverage-Report-Client-Tests/app/shared/sidebar/sidebar-card.directive.ts.html) | 100% 
| [sidebar-card-large.component](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6554-22/artifact/shared/Coverage-Report-Client-Tests/app/shared/sidebar/sidebar-card-large/sidebar-card-large.component.ts.html) | 100% 
| [sidebar-card-medium.component](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6554-22/artifact/shared/Coverage-Report-Client-Tests/app/shared/sidebar/sidebar-card-medium/sidebar-card-medium.component.ts.html) | 100% 
| [sidebar-card-small.component](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6554-22/artifact/shared/Coverage-Report-Client-Tests/app/shared/sidebar/sidebar-card-small/sidebar-card-small.component.ts.html) | 86.66% 

### Screenshots

Small Sidebar Card Example (From a draft branch)

Screenshot 2024-05-18 at 12 46 11

Medium Sidebar Card Example (Currently used, default size)

Screenshot 2024-05-18 at 12 46 17

Large Sidebar Card Example (From a draft branch)

Screenshot 2024-05-18 at 13 00 38

### CODE REVIEW

File: src/main/webapp/app/shared/sidebar/sidebar-accordion/sidebar-accordion.component.ts
Comment (rabeatwork) [inline src/main/webapp/app/shared/sidebar/sidebar-accordion/sidebar-accordion.component.ts:78]: This information should be part in the sidebar data itself. :)

File: src/main/webapp/app/shared/sidebar/sidebar-card-medium/sidebar-card-medium.component.scss
Comment (rabeatwork) [inline src/main/webapp/app/shared/sidebar/sidebar-card-medium/sidebar-card-medium.component.scss:1]: I would suggest to do a follow-up PR to combine the classes we need for all SidebarCard sizes into a separate SCSS file. This way, we won't have to duplicate them.

File: src/main/webapp/app/shared/sidebar/sidebar-card-medium/sidebar-card-medium.component.ts
Comment (JohannesWt) [inline src/main/webapp/app/shared/sidebar/sidebar-card-medium/sidebar-card-medium.component.ts:14]: I would set the sidebarItem required as it must not be null anyways.
```suggestion
    @Input({required:true}) sidebarItem: SidebarCardElement;
```

File: src/main/webapp/app/shared/sidebar/sidebar-card-small/sidebar-card-small.component.ts
Comment (JohannesWt) [inline src/main/webapp/app/shared/sidebar/sidebar-card-small/sidebar-card-small.component.ts:14]: The sidebar item should also be required that it is not undefined `@Input({required:true})`

Comment (JohannesWt) [reply src/main/webapp/app/shared/sidebar/sidebar-card-small/sidebar-card-small.component.ts:14]: Then also the question marks of force Reload become obsolete

File: src/main/webapp/app/types/sidebar.ts
Comment (JohannesWt) [inline src/main/webapp/app/types/sidebar.ts:66]: If only S, M, L are allowed as Sizes one could also Create a `SidebarCardElementSize`-Enum

**Review by JohannesWt — CHANGES_REQUESTED** [review]
Looks good in general. But you could make more use of the required Input option to make null/undefined checks obsolete

File: src/main/webapp/app/shared/sidebar/sidebar-card.directive.ts
Comment (rabeatwork) [inline src/main/webapp/app/shared/sidebar/sidebar-card.directive.ts:12]: default M

**Review by JohannesWt — APPROVED** [review]
Code looks good after changes

**Review by az108 — APPROVED** [review]
Code LGTM

**Review by pzdr7 — APPROVED** [review]
Thanks for implementing the suggested changes! While I would have preferred an enum for the card size (like e.g. ButtonSize) I suppose the type definition is fine.

Code LGTM

**Review by sarpsahinalp — APPROVED** [review]
code lgtm

File: src/main/webapp/app/shared/sidebar/sidebar-card.directive.ts
Comment (egekurt123) [reply src/main/webapp/app/shared/sidebar/sidebar-card.directive.ts:12]: will be pushed in a follow-up PR

**Review by rabeatwork — APPROVED** [review]
Maintainer approved

**Review by rabeatwork — APPROVED** [review]
re-approve

**Review by pzdr7 — APPROVED** [review]
Reapprove after merge conflicts

**Review by JohannesWt — APPROVED** [review]
Reapprove

**Review by az108 — APPROVED** [review]
reapprove

**Pull request merged**
